README.md
# C2 Alignment Calculator

The Alignment Calculator is a C2 VALAWAI component that computes the alignment
if a set of norms or (*normative system*) with respect to a value of interest
using simulation and sampling.

This component provides the functionality to compute the Shapley value of an
individual norm $n_i$ within a normative system $N$ (where $n_i \in N$), with
respect to a given value $v$. Given a normative system $N$ and a value $v$, it
is possible to quantify the degree of alignment $\mathsf{Algn}_{N,v}$ that the
norms in $N$ have with respect to value $v$ (see the [Alignment Calculator
component](https://github.com/VALAWAI/C2_alignment_calculator)). The Shapley
value $\phi_i(v)$ of norm $n_i \in N$ quantifies the contribution that the
    specific norm $n_i$ makes to that alignment. It is defined as:
    
$$
    \phi_{i}(v) = \sum\limits_{N'\subseteq N\setminus \{n_i\}}
    \frac{|N'|!\left(|N|-|N'|-1\right)!}{|N|!} \cdot \left(
\mathsf{Algn}_{N'\cup \{n_i\}, v} - \mathsf{Algn}_{N', v} \right)
$$

where the summation is taken over all the subsets of $N$
where $n_i$ is absent, $N \setminus \{n_i\}$. In general, given a normative
system $N$ and a set of norm $\{n_j, n_k, ...\} \subseteq N$, the removal of
$\{n_j, n_k, ...\}$ from $N$ is represented by a normative system $N' = N
\setminus \{n_j, n_k, ...\}$ where the values of the normative parameters tied
to $\{n_j, n_k, ...\}$ are substituted by their *baseline* quantities.
These baseline normative parameters are drawn from a *baseline normative
system* $N_{bsl}$. This baseline normative system has the same norms as $N$,
tied to the same normative parameters, but the values that these baseline
normative parameters are selected in such a way that, when it is implemented on
the model, it does not change its current state.

# Summary

 - Type: C2
 - Name: Shapley Calculator
 - Version: 1.0.0 (September 27, 2023)
 - API: [1.0.0 (February 3, 2023)](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/VALAWAI/C2_shapley_calculator/main/component-api.yml)
 - VALAWAI API: [0.1.0 (September 18, 2023)](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/VALAWAI/MOV/main/valawai-api.yml)
 - Developed by: IIIA-CSIC
 - License: [MIT](LICENSE)

# Usage

A Shapley Calculator is initialized similarly to the [Alignment
Calculator](https://github.com/VALAWAI/C2_alignment_calculator), by providing
(i) a representation of the model or system being examined (i.e. the entity upon
which norms apply to), and (ii) the semantics function $f_v$ of the value of
interest $v$ whose alignment is computed by the component. The reader is
directed to the instruction of the Alignment Calculator to understand how to
define (i) and (ii). You can use the `template.py` script as a blueprint for
developing the Shapley Calculator with your own model and values.

The Shapley Calculator component is implemented as a
[Flask](https://flask.palletsprojects.com/en/2.3.x/) application. To initialize
it, use the `create_app` function:

```python
from app import create_app

app = create_app(
    YourModel,                      # your model class
    [...],                          # your model initialization arguments
    {...},                          # your model initialization keyword arguments
    baseline_norms,                 # your baseline norms dictionary
    norms,                          # your norms dictionary
    your_value_semantics_function,  # your value semantics function
    # path_length=10,               # change if needed, default is 10
    # path_sample=500               # change if needed, default is 500
)
```

This component communicates through the following HTTP requests:

* Data messages:

    - GET `/shapley` -d norm_id

* Control messages:

    - PATCH `/bsl_norms` changes the baseline normative system
    - PATCH `/norms` changes the normative system
    - PATCH `/path_length` changes the length of the paths used to compute the
      alignment
    - PATH `/path_sample` changes the number of paths sampled to compute the
      alignment

# Deployment

Clone this repository and develop your model and value semantics functions
following the blueprint in `template.py`:

```bash
$ git clone https://github.com/VALAWAI/C2_shapley_calculator.git
```

Build your Docker image in your directory of the component repository:

```bash
$ cd /path/to/c2_shapley_calculator
$ docker build -t c2_shapley_calculator .
```

Run a Docker container with your C2 Shapley Calculator:

```bash
$ docker run --rm -d \
  --network valawai \
  --name c2_shapley_calculator \
  --mount type=bind,src="$(pwd)",target=/app \
  -p 5432:5000 \
  -e MODEL=my_model \
  c2_shapley_calculator
```

The environment variable `MODEL` refers to the script where you have defined
your model (do not include the .py extension).

Once the container is up and running, use `curl` to communicate with the
component:

```bash
$ curl -X GET http://localhost:5432/algn -d pay
{
  "shapley": 0.5502168612232454
}
```

```bash
$ curl -X PATCH http://localhost:5432/path_sample -d 200
{}
```


CHANGELOG.md
# C2_shapley_calculator
 
## Version 1.0.0 (October 3, 2023)

- Initial component version.


example.py
import argparse
from mesa import Agent, Model
from mesa.time import BaseScheduler
from multiprocessing import Process
from random import seed, uniform

from app import create_app

parser = argparse.ArgumentParser()
parser.add_argument('port', type=int)

seed(100)

class TaxModel(Model):
    def __init__(self) -> None:
        """An example Agent-Based Model where agents contribute to a common fund
        with some taxes and get some payback, according to their ranking in
        wealth.
        """
        super().__init__()
        self.num_agents = 200
        self.num_segments = 5
        self.schedule = BaseScheduler(self)
        for i in range(self.num_agents):
            a = Agent(i, self)
            a.wealth = uniform(0, 100)
            self.schedule.add(a)
        self.sorted_agents = sorted(self.schedule.agents, key=lambda a: a.wealth)
        self.ags_per_seg = self.num_agents//self.num_segments
        self.segments = [i//self.ags_per_seg for i in range(self.num_agents)]

    def step(self, norms) -> None:
        common_fund = 0.
        for i in range(self.num_agents):
            a = self.sorted_agents[i]
            seg = self.segments[i]
            pay = a.wealth * norms['pay']['rates'][seg]
            if pay > a.wealth:
                pay = a.wealth
            a.wealth -= pay
            common_fund += pay
        common_fund *= 1.05
        segs_payback = [
            common_fund * norms['payback']['rates'][seg] / self.ags_per_seg
            for seg in range(self.num_segments)
        ]
        for i in range(self.num_agents):
            a = self.sorted_agents[i]
            seg = self.segments[i]
            a.wealth += segs_payback[seg]
        self.sorted_agents = sorted(self.schedule.agents, key=lambda a: a.wealth)


def ratio_wealth_value(mdl: TaxModel) -> float:
    """Semantics function for value equality based on the ratio between the
    maximum and the minimum wealth.

    Parameters
    ----------
    mdl : TaxModel

    Returns
    -------
    float
    """
    a = mdl.sorted_agents[99].wealth
    b = mdl.sorted_agents[199].wealth
    return a/b


def gini_index_value(mdl: TaxModel) -> float:
    """Semantics function for value equality based on the Gini Index.

    Parameters
    ----------
    mdl : TaxModel

    Returns
    -------
    float
    """
    sorted_wealth = [ag.wealth for ag in mdl.sorted_agents]
    N = mdl.num_agents
    B = sum(xi * (N - i) for i, xi in enumerate(sorted_wealth)) 
    GI =  1 + (1 / N) - 2 * B / (N * sum(sorted_wealth))
    return 1 - 2*GI

        
if __name__ == '__main__':
    
    baseline_norms = {
        'pay': {'rates': [0.]*5},
        'payback': {'rates': [0.]*5}
    }

    normative_system = {
        'pay': {'rates': [0.1, 0.2, 0.3, 0.4, 0.5]},
        'payback': {'rates': [0.2, 0.2, 0.2, 0.2, 0.2]}
    }

    app = create_app(
        TaxModel,
        [],
        {},
        baseline_norms,
        normative_system,
        gini_index_value
    )

    args = parser.parse_args()
    app.run(debug=True, host='0.0.0.0', port=args.port)
    

template.py
import argparse
from typing import Any

from app import create_app

parser = argparse.ArgumentParser()
parser.add_argument('port', type=int)


class YourModel:
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        """A model whose evolution is determined by a set of norms."""
        super().__init__(*args, **kwargs)

    def step(
        self,
        norms: dict[str, dict[str, Any]]
    ) -> None:
        """The model evolves according to the set of norms in place.

        By having the norms as an input to the model's ``step()`` method, it is
        possible to model changes in the implemented normative system while the
        model is still running.

        Parameters
        ----------
        norms : Dict[Any, Dict[Any, Any]]
            A map of the norms governing the evolution of the model at that
            step. This map is provided as a dictionary from norm identifiers
            (keys) to a dictionary mapping each of the norms' parameters to
            their values.

        Raises
        ------
        NotImplementedError
        """
        raise NotImplementedError
    

def your_value_semantics_function(mdl: YourModel) -> float:
    """Value semantics function.

    Parameters
    ----------
    mdl : YourModel

    Returns
    -------
    float
        The degree of respect for the value at the current state of the model.

    Raises
    ------
    NotImplementedError
    """
    raise NotImplementedError


if __name__ == '__main__':

    baseline_norms = {
        ...
    }

    norms = {
        ...
    }

    app = create_app(
        YourModel,                      # your model class
        [...],                          # your model initialization arguments
        {...},                          # your model initialization keyword arguments
        baseline_norms,                 # your baseline norms dictionary
        norms,                          # your norms dictionary
        your_value_semantics_function,  # your value semantics function
        # path_length=10,               # change if needed, default is 10
        # path_sample=500               # change if needed, default is 500
    )

    args = parser.parse_args()
    app.run(debug=True, host='0.0.0.0', port=args.port)
    

shapley_calculator.py
from copy import deepcopy
from flask import Flask, Request, request
from itertools import combinations
from math import factorial
from pathos.multiprocessing import cpu_count, ProcessPool
from traceback import format_exc

from typing import Any, Callable, Type


def evaluate_path(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    norms: dict[str, dict[str, Any]],
    value: Callable[[object], float],
    path_length: int
) -> float:
    """Evaluate the outcome of a path in terms of a value.

    Parameters
    ----------
    model_cls : Type[...]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initialization keyword arguments.
    norms : dict[str, dict[str, Any]]
        The set of norms governing the evolution of the model.
    value : Callable[[object], float]
        The value semantics function that evaluates the final state of a path.
    path_length : int
        The number of steps in the path to evaluate.

    Returns
    -------
    float
    """
    mdl = model_cls(*model_args, **model_kwargs)
    for _ in range(path_length):
        mdl.step(norms)
    return value(mdl)


def alignment(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    norms: dict[str, dict[str, Any]],
    value: Callable[[object], float],
    path_length: int,
    path_sample: int,
    pool: ProcessPool
) -> float:
    """Compute the alignment from a sample of paths.

    This function uses a ``pathos.multiprocessing.ProcessPool`` already
    initialized to speed up the sampling.

    Parameters
    ----------
    model_cls : Type[Model]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initialization keyword arguments.
    norms : dict[str, dict[str, Any]]
        The set of norms governing the evolution of the model.
    value : Callable[[object], float]
        The value semantics function that evaluates the final state of a path.
    path_length : int
        The number of steps in the path to evaluate.
    path_sample : int
        The number of paths to sample.
    pool : ProcessPool
        A ``pathos.multiprocessing.ProcessPool`` to parallelize sampling.

    Returns
    -------
    float
    """
    args = [
        [model_cls] * path_sample,
        [model_args] * path_sample,
        [model_kwargs] * path_sample,
        [norms] * path_sample,
        [value] * path_sample,
        [path_length] * path_sample
    ]
    pool.restart()
    algn = 0.
    for res in pool.map(evaluate_path, *args):
        algn += res
    pool.close()
    pool.terminate()
    return algn / path_sample
    

def shapley_value(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    baseline_norms: dict[str, dict[str, Any]],
    normative_system: dict[str, dict[str, Any]],
    norm: str,
    value: Callable[[object], float],
    path_length: int = 10,
    path_sample: int = 100
) -> float:
    """Compute the Shapley value of a norm in a normative system.

    This calculator computes the Shapley value of individual norms in a
    normative system with respect to some value. For a complete
    formalization of the Shapley values of norms in a normative system, see
    [1]_.

    Parameters
    ----------
    model_cls : Type[...]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initialization keyword arguments.
    baseline_norms : dict[str, dict[str, Any]]
        The baseline norms causing no evolution of the ABM.
    normative_system : dict[str, dict[str, Any]]
        The complete normative system, as a mapping of norm IDs (the keys)
        to a dictionary of its normative parameters.
    norm : str
        The ID of the norm whose Shapley value is computed.
    value : Callable[[object], float]
        The value with respect to which the Shapley value is computed. It is
        passed as a function taking as input a model instance and returning
        the evaluation of the value semantics function given the state of
        the model.
    path_length : int, optional
        The length of the evolution path used to compute the alignment, by
        default 10.
    path_sample : int, optional
        The number of paths to sample when computing the alignment, by
        default 100.

    Returns
    -------
    float
        The Shapley value of ``norm`` in ``normative_system`` with respect
        to ``value``.

    References
    ----------
    .. [1] Montes, N., & Sierra, C. (2022). Synthesis and properties of
        optimally value-aligned normative systems. Journal of Artificial
        Intelligence Research, 74, 1739–1774.
        https://doi.org/10.1613/jair.1.13487
    """
    # check that norms match
    assert normative_system.keys() == baseline_norms.keys(), \
        "normative system must have identical norms to the baseline normative system"
    for norm_id, params in normative_system.items():
        assert params.keys() == baseline_norms[norm_id].keys(), \
        f"norm {norm_id} does not have the same params in the normative \
            system and in the baseline normative system"
        
    # prepare list of norms id's for the subsets of norms
    all_norms_ids = list(baseline_norms.keys())
    all_norms_except_n = deepcopy(all_norms_ids)
    all_norms_except_n.remove(norm)

    N = len(all_norms_ids)
    combo_max_size = len(all_norms_except_n)
    shapley = 0.

    # prepare pool for parallelization
    num_nodes = cpu_count()
    if path_sample <= num_nodes:
        num_nodes = path_sample
    pool = ProcessPool(nodes=num_nodes)

    for N_prime in range(combo_max_size + 1):
        factor = factorial(N_prime) * factorial(N - N_prime - 1) / factorial(N)

        for norms_prime in combinations(all_norms_except_n, N_prime):
            normative_system_prime = {}
            for n_id in all_norms_ids:
                if n_id in norms_prime:
                    normative_system_prime[n_id] = normative_system[n_id]
                else:
                    normative_system_prime[n_id] = baseline_norms[n_id]

            normative_system_prime_union = deepcopy(normative_system_prime)
            normative_system_prime_union[norm] = normative_system[norm]

            algn1 = alignment(
                model_cls, model_args, model_kwargs,
                normative_system_prime_union, value, path_length,
                path_sample, pool
            )
            algn2 = alignment(
                model_cls, model_args, model_kwargs,
                normative_system_prime, value, path_length, 
                path_sample, pool
            )

            shapley += factor*(algn1 - algn2)

    pool.terminate()
    return shapley


def create_app(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    value: Callable[[object], float]
) -> Flask:
    """Create a Flask app that computes the Shapley value.

    This C2 component of the VALAWAI architecture computes the Shapley value of
    an individual norm (within a normative system) with respect to a value [1]_.

    Parameters
    ----------
    model_cls : Type[...]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initilization keyword arguments.
    value : Callable[[object], float]
        The value with respect to which the Shapley value is computed. It is
        passed as a function taking as input a model instance and returning
        the evaluation of the value semantics function given the state of
        the model.

    Returns
    -------
    Flask
        A Flask application that can process GET /shapley requests.

    References
    ----------
    .. [1] Montes, N., & Sierra, C. (2022). Synthesis and properties of optimally
        value-aligned normative systems. Journal of Artificial Intelligence
        Research, 74, 1739–1774. https://doi.org/10.1613/jair.1. 13487
    """
    __EXPECTED_TYPES = {
        'baseline_norms': dict, 'normative_system': dict, 'norm': str,
        'path_length': int, 'path_sample': int
    }
    
    app = Flask(__name__)

    def __check_request(request: Request):
        if not request.is_json:
            return {"error": "Request must be JSON"}, 415
        input_data = request.get_json()
        if not isinstance(input_data, dict):
            return {"error": f"Params must be passed as a dict"}, 400
        return input_data


    @app.get('/shapley')
    def get_shapley():
        input_data = __check_request(request)

        __NEED_KEYS = ['baseline_norms', 'normative_system', 'norm']
        __OPT_KEYS = ['path_length', 'path_sample']

        # check that the input data has all the necessary keys and they are the
        # correct type
        for k in __NEED_KEYS:
            if not k in input_data.keys():
                return {"error": f"missing necessary param {k}"}, 400
        for k in input_data.keys():
            if not isinstance(input_data[k], __EXPECTED_TYPES[k]):
                return {"error": f"type of param {k} must be {__EXPECTED_TYPES[k]}"}, 400

        # get optional arguments path_length and path_sample
        kwargs = {}
        for k in __OPT_KEYS:
            try:
                kwargs[k] = input_data[k]
            except KeyError:
                continue

        # compute and return
        try:
            shap = shapley_value(
                model_cls,
                model_args,
                model_kwargs,
                input_data['baseline_norms'],
                input_data['normative_system'],
                input_data['norm'],
                value,
                **kwargs
            )
            return {'shapley': shap}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    return app


app.py
# Copyright 2023 IIIA-CSIC
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License version 3
# along with this program; if not, see <http://www.gnu.org/licenses/>.

from dataclasses import dataclass
from flask import Flask, Request, request, escape
from traceback import format_exc
from valalgn.sampling import shapley_value

from typing import Any, Callable, Type

@dataclass
class ShapleyCalculator:
    model_cls: Type
    model_args: list[Any]
    model_kwargs: dict[str, Any]
    baseline_norms: dict[str, dict[str, Any]]
    norms: dict[str, dict[str, Any]]
    value: Callable[[object], float]
    path_length: int = 10
    path_sample: int = 500

    def compute_shapley_value(self, norm: str) -> float:
        return shapley_value(
            self.model_cls,
            self.model_args,
            self.model_kwargs,
            self.baseline_norms,
            self.norms,
            norm,
            self.value,
            self.path_length,
            self.path_sample
        )

def create_app(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    baseline_norms: dict[str, dict[str, Any]],
    norms: dict[str, dict[str, Any]],
    value: Callable[[object], float],
    path_length: int = 10,
    path_sample: int = 500
) -> Flask:
    """Create a Flask app that computes the Shapley value of a norm.

    This C2 component of the VALAWAI architecture computes the Shapley value of
    an individial in a normative system, with respect to a value (or an
    aggregation of values) [1]_.

    Parameters
    ----------
    model_cls : Type
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initilization keyword arguments.
    baseline_norms : dict[str, dict[str, Any]]
        The set of baseline norms, whose implementation would not cause the
        model to evolve.
    norms : dict[str, dict[str, Any]]
        The set of norms governing the evolution of the model.
    value : Callable[[object], float]
        The value with respect to which the alignment is computed. It is passed
        as a function taking as input a model instance and returning the
        evaluation of the value semantics function given the state of the model.
    path_length : int, optional
        The length of the paths to evaluate the alignment, by default 10.
    path_sample : int, optional
        The sample size of the paths to evaluate the alignment, by default 500.

    Returns
    -------
    Flask
        A Flask application that can process GET /shapley/{norm_id} requests.

    References
    ----------
    .. [1] Montes, N., & Sierra, C. (2022). Synthesis and properties of
        optimally value-aligned normative systems. Journal of Artificial
        Intelligence Research, 74, 1739–1774. https://doi.org/10.1613/jair.1.
        13487
    """
    app = Flask(__name__)
    
    calculator = ShapleyCalculator(
        model_cls, model_args, model_kwargs, baseline_norms, norms, value, path_length, path_sample
    )

    def __check_request(request: Request):
        if not request.is_json:
            return {"error": "Request must be JSON"}, 415
        input_data = request.get_json()
        if not isinstance(input_data, dict):
            return {"error": f"Params must be passed as a dict"}, 400
        return input_data

    @app.get('/shapley')
    def get_shapley():
        try:
            input_data = request.get_data()
            shap = calculator.compute_shapley_value(input_data.decode())
            return {'shapley': shap}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/bsl_norms')
    def patch_bsl_norms():
        try:
            input_data = __check_request(request)
            for k, v in input_data.items():
                calculator.baseline_norms[k].update(v)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/norms')
    def patch_norms():
        try:
            input_data = __check_request(request)
            for k, v in input_data.items():
                calculator.norms[k].update(v)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/path_length')
    def patch_path_length():
        try:
            input_data = request.get_data()
            calculator.path_length = int(input_data)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/path_sample')
    def patch_path_sample():
        try:
            input_data = request.get_data()
            calculator.path_sample = int(input_data)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    return app